bsp/Src/I2C.c

```diff
@@ -0,0 +1,442 @@
+	update_last_i2c(hi2c);
+
+	if(HAL_I2C_GetState(hi2c) == HAL_I2C_STATE_READY)
+	{
+		return I2C_ConvertHALStatus(HAL_I2C_Master_Transmit_IT(hi2c, deviceAddr, pDataBuff, len));
+	}
+	
+	QueueHandle_t I2C_Queue = I2C1_Queue;
+	if(hi2c->Instance == I2C1) {}
+	else if (hi2c->Instance == I2C2)
+	{
+		I2C_Queue = I2C2_Queue;
+	}
+	else if (hi2c->Instance == I2C3)
+	{
+		I2C_Queue = I2C3_Queue;
+	}
+	else
+	{
+		// Handle invalid I2C peripheral
+		return I2C_BADINST;
+	}
+
+	if (len > uxQueueSpacesAvailable(I2C_Queue))
+	{
+		return I2C_QFULL;
+	}
+	
+	for(int packetsLeft = len - 1; packetsLeft >= 0; packetsLeft--)
+	{
+		DataInfo_t dataToEnqueue = {
+			.hi2c = hi2c,
+			.deviceAddr = deviceAddr,
+			.pDataBuff = pDataBuff[len - packetsLeft - 1],
+			.payloadsLeft = packetsLeft
+		};
+
+		if (xQueueSend(I2C_Queue, &dataToEnqueue, 0) != pdPASS)
+		{
+			// Handle queue full error
+			return I2C_QFULL;
+		}
+	}
+
+	return I2C_SENT;
+}
+
+/**
+* @brief    Receives data from the I2C bus.
+* @param    hi2c:            to structure that has config data
+* @param    deviceAddr      target device address
+* @param    pDataBuff      data buffer
+* @param    len            amount of data
+* @return   I2C_StatusTypeDef
+*/
+I2C_StatusTypeDef i2c_recv(I2C_HandleTypeDef *hi2c,
+              uint8_t deviceAddr,
+              uint8_t* pDataBuff,
+              uint16_t len)
+{
+	// Update the latest I2C peripheral
+	update_last_i2c(hi2c);
+
+    return I2C_ConvertHALStatus(HAL_I2C_Master_Receive_IT(hi2c, deviceAddr, pDataBuff, len));
+}
+
+/**
+ * @brief 	Internal function to send the value of the specified i2c
+ * peripheral's queue
+ *
+ * @param	i2c_peripheral Either I2C1, I2C1, or I2C3
+ */
+static void queue_send(I2C_TypeDef *i2c_peripheral)
+{
+	I2C_HandleTypeDef *last_hi2c = NULL;
+	QueueHandle_t *I2C_Queue = NULL;
+	BaseType_t higherPriorityTaskWoken = pdFALSE;
+
+	#ifdef I2C1
+	if (i2c_peripheral == I2C1)
+	{
+		last_hi2c = last_hi2c1;
+		I2C_Queue = &I2C1_Queue;
+	}
+	#endif
+	#ifdef I2C2
+	if (i2c_peripheral == I2C2)
+	{
+		last_hi2c = last_hi2c2;
+		I2C_Queue = &I2C2_Queue;
+	}
+	#endif
+	#ifdef I2C3
+	if (i2c_peripheral == I2C3)
+	{
+		last_hi2c = last_hi2c3;
+		I2C_Queue = &I2C3_Queue;
+	}
+	#endif
+
+	if(I2C_Queue == NULL || last_hi2c == NULL)
+	{
+		// Invalid I2C peripheral
+		return;
+	}
+
+	if (HAL_I2C_GetState(last_hi2c) != HAL_I2C_STATE_READY)
+		return;
+
+	DataInfo_t data;
+	if (xQueueReceiveFromISR(*I2C_Queue, &data, &higherPriorityTaskWoken) == pdPASS) {
+	    // Process dequeuedData
+		if(i2c_peripheral == I2C1)
+		{
+			uint8_t dataToTransmit[I2C1_MAX_PAYLOADS];
+			uint8_t index = 0;
+			dataToTransmit[index] = data.pDataBuff;
+			while(data.payloadsLeft != 0)
+			{
+				xQueueReceiveFromISR(*I2C_Queue, &data, &higherPriorityTaskWoken);
+				dataToTransmit[++index] = data.pDataBuff;
+			}
+
+			HAL_I2C_Master_Transmit_IT(data.hi2c, data.deviceAddr, dataToTransmit, index + 1);	
+		}
+		else if(i2c_peripheral == I2C2)

```

Comment:
i hate spamming ifdefs but for uniformity it makes sense to put them here for `I2C2` and `I2C3`

Reply:
I've gone ahead and decided to have the implication that I2C1 will always be defined because the header should only be defined if I2C1 is defined. This is under the assumption that the existence of I2C3 implies the existence of I2C2 same logic for I2C2 implying I2C1.

Reply:
eh. what if the user includes the header on a board that doesn't support I2C and tries running code. an extra check don't hurt.

Reply:
The header shouldn't compile if the board doesn't have at least I2C1 defined, so it is unneeded.

Reply:
There are other ways that a user could misuse the function without including the header, including simply defining the method stub in their current file (of course, they shouldn't be doing this, but it is technically possible to get around using the header at all). Adding an I2C1 ifdef here is a simple way to idiot-proof the driver.

Reply:
As stated in a couple of other comments I have put on here. I would like to move forward without having the I2C1 as it would match the style in the CAN bsp.

Reply:
Alright fine